File: lcode/ex30/isSymmetric.go
// Copyright 2019 readailib.com. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package ex30

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	if root==nil{
		return right==nil
	}

	return isChildSymmetric(root.Left, root.Right)
}

func isChildSymmetric(left, right *TreeNode) bool {
	if left == nil {
		return true
	}

	if right == nil {
		return false
	}

	return left.Val == right.Val && isChildSymmetric(left.Left, right.Right) && isChildSymmetric(left.Right, right.Left) 
}

File: lcode/ex30/isSymmetric.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        return self.isChildSymmetric(root.left, root.right)
        
    def isChildSymmetric(self, left, right):
        if left is None:
            return right == None

        # 这个就多余了
        if right is None:
            return left == None

        return left.val == right.val and\
             self.isChildSymmetric(left.right, right.left) and \
             self.isChildSymmetric(left.left, right.right)

        
